File: backend/update_schema.js
const fs = require('fs');
const path = require('path');

const schemaPath = path.join(__dirname, 'prisma', 'schema.prisma');
let schema = fs.readFileSync(schemaPath, 'utf8');

// Replace function to add indexes before @@map
function addIndexBeforeMap(modelMapName, indexStrings) {
  const mapStr = `@@map("${modelMapName}")`;
  const indexesStr = indexStrings.map(idx => `  ${idx}`).join('\n');
  
  if (schema.includes(indexesStr)) return; // Already added
  
  schema = schema.replace(
    new RegExp(`\\s*@@map\\("${modelMapName}"\\)`),
    `\n${indexesStr}\n  @@map("${modelMapName}")`
  );
}

// Aquasphere
addIndexBeforeMap('items', ['@@index([name])', '@@index([type])']);
addIndexBeforeMap('vendors', ['@@index([name])']);
addIndexBeforeMap('users', ['@@index([role])', '@@index([is_active])']);
addIndexBeforeMap('orders', ['@@index([delivery_status, payment_status])']);
addIndexBeforeMap('spot_sales', ['@@index([customer_id, created_at(sort: Desc)])']);

// Wadaana
// But wait, the Wadaana maps might be the same. 
// e.g. @@map("items") @@schema("wadaana")
// If they have the same @@map, my regex will only replace the first one.
// Let's use schema specific replacement.

File: backend/update_schema.py
import re

with open('prisma/schema.prisma', 'r') as f:
    schema = f.read()

def add_indexes(model_name, schema_name, indexes):
    global schema
    # Find the block for the model
    # Matches: model ModelName { ... @@map("table_name") @@schema("schema_name") }
    pattern = rf'model\s+{model_name}\s+{{[^}}]*?@@map\("([^"]+)"\)\s*@@schema\("{schema_name}"\)[^}}]*?}}'
    
    def replacer(match):
        block = match.group(0)
        # Check if already has the indexes
        if all(idx in block for idx in indexes):
            return block
        
        # Insert indexes right before @@map
        idx_str = '\n'.join(f'  {idx}' for idx in indexes)
        new_block = re.sub(rf'(\s*@@map\("[^"]+"\))', f'\n{idx_str}\\1', block)
        return new_block

    schema = re.sub(pattern, replacer, schema)

# Aquasphere
add_indexes('AquasphereItem', 'aquasphere', ['@@index([name])', '@@index([type])'])
add_indexes('AquasphereVendor', 'aquasphere', ['@@index([name])'])
add_indexes('AquasphereUser', 'aquasphere', ['@@index([role])', '@@index([isActive])'])
add_indexes('AquasphereOrder', 'aquasphere', ['@@index([deliveryStatus, paymentStatus])'])
add_indexes('AquasphereSpotSale', 'aquasphere', ['@@index([customerId, createdAt(sort: Desc)])'])

# Wadaana
add_indexes('WadaanaItem', 'wadaana', ['@@index([name])', '@@index([type])'])
add_indexes('WadaanaVendor', 'wadaana', ['@@index([name])'])
add_indexes('WadaanaOrder', 'wadaana', ['@@index([deliveryStatus, paymentStatus])'])
add_indexes('WadaanaSpotSale', 'wadaana', ['@@index([customerId, createdAt(sort: Desc)])'])
# No WadaanaUser

with open('prisma/schema.prisma', 'w') as f:
    f.write(schema)
